support/Pipelines/Bbh/ControlId.py

```diff
@@ -0,0 +1,190 @@
+                dimensionless_spin_a=[chi_Kerr_A_x, chi_Kerr_A_y, chi_Kerr_A_z],
+                dimensionless_spin_b=[chi_Kerr_B_x, chi_Kerr_B_y, chi_Kerr_B_z],
+                separation=separation,
+                orbital_angular_velocity=orbital_angular_velocity,
+                radial_expansion_velocity=radial_expansion_velocity,
+                run_dir=control_run_dir,
+                control=False,
+                evolve=False,
+                scheduler=None,
+            )
+
+        # Get black hole physical parameters
+        with spectre_h5.H5File(
+            f"{control_run_dir}/Horizons.h5", "r"
+        ) as horizons_file:
+            AhA_quantities = to_dataframe(
+                horizons_file.get_dat("AhA.dat")
+            ).iloc[-1]
+
+            M_A = AhA_quantities["ChristodoulouMass"]
+            chi_A_x = AhA_quantities["DimensionlessSpinVector_x"]
+            chi_A_y = AhA_quantities["DimensionlessSpinVector_y"]
+            chi_A_z = AhA_quantities["DimensionlessSpinVector_z"]
+
+            horizons_file.close_current_object()
+            AhB_quantities = to_dataframe(
+                horizons_file.get_dat("AhB.dat")
+            ).iloc[-1]
+
+            M_B = AhB_quantities["ChristodoulouMass"]
+            chi_B_x = AhB_quantities["DimensionlessSpinVector_x"]
+            chi_B_y = AhB_quantities["DimensionlessSpinVector_y"]
+            chi_B_z = AhB_quantities["DimensionlessSpinVector_z"]
+
+        residual = np.array(
+            [
+                M_target_A - M_A,
+                M_target_B - M_B,
+                chi_target_A[0] - chi_A_x,
+                chi_target_A[1] - chi_A_y,
+                chi_target_A[2] - chi_A_z,
+                chi_target_B[0] - chi_B_x,
+                chi_target_B[1] - chi_B_y,
+                chi_target_B[2] - chi_B_z,
+            ]
+        )
+        logger.info(f"Control Residual = {np.max(np.abs(residual)):e}")
+
+        data_file.write(
+            f" {iteration},"
+            f" {M_A}, {M_B},"
+            f" {chi_A_x}, {chi_A_y}, {chi_A_z},"
+            f" {chi_B_x}, {chi_B_y}, {chi_B_z},"
+            f" {residual[0]}, {residual[1]},"
+            f" {residual[2]}, {residual[3]}, {residual[4]},"
+            f" {residual[5]}, {residual[6]}, {residual[7]}"
+            " \n"
+        )
+        data_file.flush()
+
+        return residual
+
+    # Initial guess
+    u = np.array([M_target_A, M_target_B, *chi_target_A, *chi_target_B])
+
+    # Initial residual
+    F = Residual(*u)
+
+    # Initial Jacobian (identity matrix)
+    J = np.identity(len(u))
+
+    while iteration < max_iterations and np.max(np.abs(F)) > residual_tolerance:
+        iteration += 1
+
+        # Update the free parameters using a quasi-Newton-Raphson method
+        Delta_u = np.dot(np.linalg.inv(J), F)
+        u += Delta_u
+
+        F = Residual(*u)
+
+        # Update the Jacobian using Broyden's method
+        J = J - np.dot(F, Delta_u) / np.linalg.norm(Delta_u)
```

Comment:
In the literature I find always `np.dot(Delta_u, Delta_u)`, i.e. the squared L2 norm (e.g. Numerical Recipes). I think the Ossokine2015 paper might just be missing a square. Could you check this please?

Reply:
I've fixed the dot product. There was also an issue of a missing outer product (instead of the dot product) between `F` and `Delta_u`. Now the control loop reaches `1.e-6` in only 4 iterations (< 30min).

Reply:
Awesome!